File: SortingAlgorithm/inversionCount.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution{
public:
    int merge(vector<int> &arr, int start, int end, int mid){
        int count = 0;
        int i = 0, j = 0, k = start;
        int n = mid - start + 1;
        int m = end - mid;
        vector<int> left;
        vector<int> right;
        for(int l = start; l < mid+1; l++ ) left.push_back(arr[l]);
        for (int r = mid+1; r < end+1; r++) right.push_back(arr[r]);
        while (i < n and j < m){
            if (left[i] <= right[j]){
                arr[k++] = left[i++];
            }
            else{
                count += n - i;
                arr[k++] = right[j++];
            }
        }
        while (i < n) arr[k++] = left[i++];
        while (j < m) arr[k++] = right[j++];
        return count;

    }

    int mergeSort(vector<int> &arr, int start,int end){
        int inv = 0;
        if (start < end){
            int mid = start + (end - start)/2;
            inv += mergeSort(arr, start, mid);
            inv += mergeSort(arr, mid+1, end);
            inv+= merge(arr, start, end, mid);

        }
        return inv;
    }
};

File: SortingAlgorithm/inversionCount.py
class Solution:
    # User function Template for python3

    # arr[]: Input Array
    # N : Size of the Array arr[]
    # Function to count inversions in the array.
    def merge(self, arr, start, end, mid):
        count = 0
        n = mid - start + 1
        m = end - mid
        left = []
        right = []
        for i in range(start, mid+1):
            left.append(arr[i])
        for i in range(mid+1, end+1):
            right.append(arr[i])
        i = 0
        j = 0
        k = start
        while i < n and j < m:
            if left[i] <= right[j]:
                arr[k] = left[i]
                k += 1
                i += 1
            else:
                count += n - i
                arr[k] = right[j]
                j += 1
                k += 1
        while i < n:
            arr[k] = left[i]
            i += 1
            k += 1
        while j < m:
            arr[k] = right[j]
            j += 1
            k += 1
        return count

    def mergeSort(self, arr, start, end):
        inv = 0
        if start < end:
            mid = start + (end - start) // 2
            inv += self.mergeSort(arr, start, mid)
            inv += self.mergeSort(arr, mid + 1, end)
            inv += self.merge(arr, start, end, mid)
        return inv

    def inversionCount(self, arr, n):
        return self.mergeSort(arr, 0, n-1)


# Your Code Here

obj = Solution()
print(obj.inversionCount([2, 4, 1, 3, 5], 5))
# print(obj.merge([1, 2, 4, 6, 3, 5, 7, 8], 0, 8, 4))
